### PR TITLE
enh(api): finish implementation for renew password use case

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

this PR intends to finish implementation for renew password use case

**Fixes** # MON-12290

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

## Checklist

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Core/Application/Security/Exception/UserPasswordException.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?php
+
```
Comment (kduret) [inline src/Core/Application/Security/Exception/UserPasswordException.php:2]: missing license

File: src/Centreon/Domain/Common/Assertion/Assertion.php [hunk 2]
```diff
@@ -283,4 +283,28 @@ function (array $parameters) {
...
+
+    /**
+     * Assert that a value match a regex
+     *
+     * @param mixed $value
+     * @param string $pattern
+     * @param string|null $propertyPath
+     * @return void
+     */
+    public static function regex($value, string $pattern, string $propertyPath = null): void
```
Comment (kduret) [inline src/Centreon/Domain/Common/Assertion/Assertion.php:295]:
```suggestion
    public static function regex(mixed $value, string $pattern, string $propertyPath = null): void
```

File: src/Core/Domain/Security/User/Model/UserPasswordFactory.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
+<?php
+
```
Comment (kduret) [inline src/Core/Domain/Security/User/Model/UserPasswordFactory.php:2]: missing license

File: src/Core/Domain/Security/User/Model/UserPasswordFactory.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            if ($securityPolicy->hasSpecialCharacter()) {
+                Assertion::regex(
+                    $password,
+                    '/[' . Configuration::SPECIAL_CHARACTERS_LIST . ']/',
+                    'UserPassword::passwordValue'
+                );
+            }
+        } catch (AssertionException $ex) {
+            //Throw a generic user password exception to avoid returning a plain password in the AssertionException.
+            throw UserPasswordException::passwordDoesntMatchSecurityPolicy();
```
Comment (kduret) [inline src/Core/Domain/Security/User/Model/UserPasswordFactory.php:50]: I think it's better to not use shorthand in method naming
```suggestion
            throw UserPasswordException::passwordDoesnotMatchSecurityPolicy();
```

**Review by kduret — APPROVED** [review]

**Review by AssanDiomande — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**